### PR TITLE
Revision test

### PR DESCRIPTION
**Pull request by maririos**
Changes include:
- Generate a different setting per test
- Add Revisions test `/revisions`
- Changes in `ConfigurationSettingParser`

### CODE REVIEW

File: Configuration/Azure.Configuration.Test/ConfigurationLiveTests.cs [hunk 1]
```diff
@@ -186,6 +201,55 @@ public async Task UpdateIfAny()
...
+            await service.SetAsync(testSetting, CancellationToken.None);
+            await service.SetAsync(testSettingUpdate, CancellationToken.None);
+
+            try
+            {
+                // Test
+                var filter = new BatchFilter();
+                filter.Key = testSetting.Key;
+                
+                Response<SettingBatch> response = await service.RevisionsAsync(filter, CancellationToken.None);
```
Comment (KrzysztofCwalina) [inline Configuration/Azure.Configuration.Test/ConfigurationLiveTests.cs]: You should check if the response status is 200

File: Configuration/Azure.Configuration/ConfigurationClient.cs [hunk 2]
```diff
@@ -236,5 +236,36 @@ public async Task<Response<SettingBatch>> GetBatchAsync(BatchFilter filter, Canc
                 return new Response<SettingBatch>(response, batch);
             }
         }
+
+        public async Task<Response<SettingBatch>> RevisionsAsync(BatchFilter filter, CancellationToken cancellation = default)
```
Comment (KrzysztofCwalina) [inline Configuration/Azure.Configuration/ConfigurationClient.cs]: Our design guidelines say that method names should be verb phrases. "Revisions" is a noun (and we use nouns for properties). So, you might want to rename it to something like GetRevisionsAsync.

File: Configuration/Azure.Configuration/ConfigurationClient.cs [hunk 3]
```diff
@@ -236,5 +236,36 @@ public async Task<Response<SettingBatch>> GetBatchAsync(BatchFilter filter, Canc
                 return new Response<SettingBatch>(response, batch);
             }
         }
+
+        public async Task<Response<SettingBatch>> RevisionsAsync(BatchFilter filter, CancellationToken cancellation = default)
+        {
+            var uri = BuildUrlForRevisions(filter);
```
Comment (KrzysztofCwalina) [inline Configuration/Azure.Configuration/ConfigurationClient.cs]: BTW, I used to have a type called Url, and so some of the helper methods are called BuildUrl. You might want to rename them to BuildUri....

File: Configuration/Azure.Configuration/ConfigurationClient.cs (hunk 2 above)
Comment (maririos) [reply Configuration/Azure.Configuration/ConfigurationClient.cs]: Could you point me to where the design guidelines are?

File: Configuration/Azure.Configuration/ConfigurationClient_private.cs [hunk 4]
```diff
@@ -119,32 +119,49 @@ Uri BuildUriForLocksRoute(string key, SettingFilter filter)
             return builder.Uri;
         }
 
-        Uri BuildUrlForGetBatch(BatchFilter options)
+        UriBuilder BuildBatchQuery(UriBuilder builder, BatchFilter options)
```
Comment (KrzysztofCwalina) [inline Configuration/Azure.Configuration/ConfigurationClient_private.cs]: Why do you both take the builder as a parameter and return the same instance?

File: Configuration/Azure.Configuration/ConfigurationClient.cs (hunk 3 above)
Comment (maririos) [reply Configuration/Azure.Configuration/ConfigurationClient.cs]: good point. Will change

File: Configuration/Azure.Configuration.Test/ConfigurationLiveTests.cs (hunk 1 above)
Comment (maririos) [reply Configuration/Azure.Configuration.Test/ConfigurationLiveTests.cs]: Thank you! I forgot to add that :|

File: Configuration/Azure.Configuration/ConfigurationSettingParser.cs [hunk 5]
```diff
@@ -164,10 +165,12 @@ static bool TryParse(ReadOnlySequence<byte> content, out List<ConfigurationSetti
...
                 {
                     case JsonTokenType.StartObject:
-                        value = new ConfigurationSetting();
+                        // TODO (pri 1): manage tag objects
+                        if(state != JsonState.tags) value = new ConfigurationSetting();
                         break;
                     case JsonTokenType.EndObject:
-                        result.Add(value);
+                        // TODO (pri 1): manage tag objects
+                        if (state != JsonState.tags) result.Add(value); ;
```
Comment (KrzysztofCwalina) [inline Configuration/Azure.Configuration/ConfigurationSettingParser.cs]: nit: two semicolons

File: Configuration/Azure.Configuration/ConfigurationClient_private.cs (hunk 4 above)
Comment (maririos) [reply Configuration/Azure.Configuration/ConfigurationClient_private.cs]: To make explicit that I'm changing something in the builder.

File: Configuration/Azure.Configuration/ConfigurationClient.cs (hunk 2 above)
Comment (KrzysztofCwalina) [reply Configuration/Azure.Configuration/ConfigurationClient.cs]: Full guidelines: https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/
Particular guidelines for naming members/methods: https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/names-of-type-members

File: Configuration/Azure.Configuration/ConfigurationClient_private.cs (hunk 4 above)
Comment (KrzysztofCwalina) [reply Configuration/Azure.Configuration/ConfigurationClient_private.cs]: I don't think it makes it more explicit. Without looking at the implementation, and just looking at the signature, people might think it creates a new instance (let's say a copy/clone) and returns it. 
We have many APIs in the Framework that mutate reference types and they don't return the mutated object. For example, all APIs that read from streams, e.g. JsonObject.Parse(Stream), mutate the input parameter.

File: Configuration/Azure.Configuration.Test/ConfigurationLiveTests.cs [hunk 6]
```diff
@@ -186,6 +201,57 @@ public async Task UpdateIfAny()
...
+            Assert.NotNull(connectionString, "Set AZ_CONFIG_CONNECTION environment variable to the connection string");
+            var service = new ConfigurationClient(connectionString);
+
+            //Prepare environment
+            ConfigurationSetting testSetting = GenerateSetting();
+            var testSettingUpdate = testSetting.Clone();
+            testSettingUpdate.Label = "test_label_update";
+            int expectedEvents = 2;
+
+            await service.SetAsync(testSetting, CancellationToken.None);
```
Comment (KrzysztofCwalina) [inline Configuration/Azure.Configuration.Test/ConfigurationLiveTests.cs]: And in general, for all the service calls, we should decide what to do with the response status. Maybe simply assert that the status equals 200.

**Comment by maririos** [comment]
All PR comments have been addressed

File: Configuration/Azure.Configuration/ConfigurationClient_private.cs [hunk 7]
```diff
@@ -123,31 +123,49 @@ Uri BuildUriForLocksRoute(string key, SettingFilter filter)
...
                 builder.AppendQuery(FieldsQueryFilter, filter);
             }
+        }
+
+        Uri BuildUriForGetBatch(SettingBatchFilter options)
+        {
+            var builder = new UriBuilder(_baseUri);
+            builder.Path = KvRoute;
+
+            BuildBatchQuery(ref builder, options);
```
Comment (KrzysztofCwalina) [inline Configuration/Azure.Configuration/ConfigurationClient_private.cs]: You don't need to pass it by-ref. You would need if it was a struct.

**Comment by maririos** [comment]
Thanks for all your feedback @KrzysztofCwalina :) !

**Pull request merged**